experimental/offline_attestation/README.md

```diff
@@ -0,0 +1,111 @@
+# Offline Attestation
+
+The current Oak Functions Runtime implementation uses an interactive handshake
+for setting up an attested end-to-end encrypted channel between the client and
+the runtime. For some scenarios it might be preferable to avoid the handshake.
+
+The offline attestation approach splits the attestation process into two
+distinct parts:
+
+- Key generation and export: A single asymmetric key-pair is generated. The
+  public key is bound to an attestation report that provides evidence about the
+  enclave that generated the private key. The public key and attestation report
+  are then made availble to clients.
+- Encrypted communications: Clients use the public key (once they are satisfied
+  by the evidence that only appropriate enclaves would have access to the
+  private key) to encrypt requests to the server. Clients also provide
+  additional information (e.g. the client's public key) in the request to allow
+  the server to encrypt the response.
+
+- This approach has some advantages compared to the interactive handshake-based
+  approach:
+  - Reduced bandwith and latency
+  - No need for sticky sessions in a load-balanced environment
+- And some disadvantages:
+  - No forward secrecy since the same public key is used for encrypting all
+    requests
+  - No built-in replay protection
+
+## Implementation Details
+
+This experimental example provides simplified client and server implementations.
+
+### Server
+
+The server generates an asymmetric key-pair at startup. It then generates an
+attestation report that binds the public key to the enclave hardware and the
+version of the server code. For this example we just generate a fake empty
+attestation report that is always accepted by the client.
+
+For more realistic implementations that support load-balancing a
+key-provisioning mechanism is required to ensure that all instances of the
+server have access to the private key. This could be done in two ways:
+
+- A separate key management server (KMS): The KMS would generate the key-pair
+  and attestation report. When a server instance starts up it sets up a
+  mutually-attested connection with the KMS and requests the private key. The
+  KMS is then responsible for validating that the server is running an
+  appropriate version of the software in an appropriate enclave before sharing
+  the private key. The attestation report will include evidence about the KMS,
+  not the server, so the KMS code would need to be reviewed to ensure that it
+  will do the server validation correctly.
+- Key-sharing between servers: The first server to start up will be instructed
+  to generate a key-pair and attestation reports. When other servers start up,
+  they will be instructed to connect to an existing server to get the private
+  key via a mutually-attested connection. This approach might present some
+  challenges for key rotation or upgrading to new versions of the server code.
```

Comment:
Why can't they connect to the KMS as well?
Is it needed for scalability?

Reply:
These problems for this option are cause when there is no KMS and the keys are just kept alive between all the server instance. If a KMS is introduced it should provide a way to work around these.